esp-config/src/generate.rs

```diff
@@ -0,0 +1,213 @@
+                    [b'0', b'o', ..] => usize::from_str_radix(&s[2..], 8),
+                    [b'0', b'b', ..] => usize::from_str_radix(&s[2..], 2),
+                    _ => usize::from_str_radix(&s, 10),
+                }
+                .map_err(|_| ParseError(format!("Invalid numerical value: {}", s)))?,
+            ),
+            Value::String(_) => Value::String(String::from(s)),
+        };
+        Ok(())
+    }
+
+    fn as_string(&self) -> String {
+        match self {
+            Value::Bool(value) => String::from(if *value { "true" } else { "false" }),
+            Value::Number(value) => format!("{}", value),
+            Value::String(value) => value.clone(),
+        }
+    }
+}
+
+/// Generate and parse config from a prefix, and array of key, default,
+/// description tuples.
+///
+/// This function will parse any `SCREAMING_SNAKE_CASE` environment variables
+/// that match the given prefix. It will then attempt to parse the [`Value`].
+/// Once the config has been parsed, this function will emit `snake_case` cfg's
+/// _without_ the prefix which can be used in the dependant crate. After that,
+/// it will create a markdown table in the `OUT_DIR` under the name
+/// `{prefix}_config_table.md` where prefix has also been converted
+/// to `snake_case`. This can be included in crate documentation to outline the
+/// available configuration options for the crate.
+///
+/// The tuple ordering for the`config` array is key, default, description.
+///
+/// Unknown keys with the supplied prefix will cause this function to panic.
+pub fn generate_config(prefix: &str, config: &[(&str, Value, &str)]) {
+    // ensure that the prefix is `SCREAMING_SNAKE_CASE`
+    let mut prefix = screaming_snake_case(prefix);
+
+    let mut doc_table = String::from(DOC_TABLE_HEADER);
+
+    // only rebuild if build.rs changed. Otherwise Cargo will rebuild if any
+    // other file changed.
+    println!("cargo:rerun-if-changed=build.rs");
+
+    // Generate the template for the config
+    let mut configs = HashMap::new();
+    for (name, default, desc) in config {
+        let name = format!("{prefix}_{}", screaming_snake_case(&name));
+        // insert into config
+        configs.insert(name.clone(), default.clone());
+
+        // write doc table line
+        let default = default.as_string();
+        writeln!(doc_table, "|**{name}**|{desc}|{default}|").unwrap();
+
+        // Rebuild if config envvar changed.
+        // Note this is currently buggy and requires a clean build - PR is pending https://github.com/rust-lang/cargo/pull/14058
+        println!("cargo:rerun-if-env-changed={name}");
+    }
+
+    let mut unknown = Vec::new();
+    let mut failed = Vec::new();
+
+    // Try and capture input from the environment
+    for (var, value) in env::vars() {
+        if let Some(name) = var.strip_prefix(&format!("{prefix}_")) {
+            let name = snake_case(name);
+            let Some(cfg) = configs.get_mut(&name) else {
```

Comment:
now `configs` contains SCREAMING_SNAKE_CASE incl. prefix while we look for SNAKE_CASE w/o prefix

Reply:
Oops! Thanks. I've added a test to check that the config system does what it's meant to do now